### DISCUSSION TITLE
How are browser versions handled in Cypress?

### DISCUSSION THREAD

**Discussion by Rutix** [cypress-io/cypress#18751] [CI setup] [answered]
Hi,

I was wondering if anyone can shed some light on the following:

We are running Cypress in Azure Devops Pipelines and to speed up the build we run multiple cloud agents in parallel. We get these agents from Microsoft and then run `npm install` on them etc. Now we have the following problem: whenever Chrome is updated (and is rolling out the new updates over several days to systems) we run into the issue that of the 8 agents, several of them will get the new version but the rest are still on the old one. So in the case of today we had 2 agents which ended up running chrome 95.0.4638.69 but the rest still had 94.0.4606.81. The result of this is that the whole pipeline will fail since one of those stages resulted in an error. The error being:

This machine is sending different environment parameters than the first machine that started this parallel run.

The existing run is: xxx

In order to run in parallel mode each machine must send identical environment parameters such as:

- specs
- osName
- osVersion
- browserName
- browserVersion (major)

This is the third time it is happening so we are now looking at a perm solution which will prevent this in the future. We tried looking at the documentation but it's not clear how Chrome is installed by Cypress and if we can control the versions (so that we can switch to the newer version when we are sure it's rolled out everywhere). Can anyone shed some light on this?

**Comment by Rutix** [comment, marked as answer]
Never mind, found out that it uses the installed Chrome version and Cypress doesn't install it itself. As information: https://github.com/actions/virtual-environments/blob/main/images/linux/Ubuntu2004-README.md the agent uses this image which probably isn't updated yet on all agents Microsoft supplies.

**Comment by kevinold** [comment]
@Rutix If you are able, please provide a Azure Pipelines YAML, but it sounds like a custom container is not being used in your configuration.

Most likely passing a [container](https://docs.microsoft.com/en-us/azure/devops/pipelines/yaml-schema?view=azure-devops&tabs=example%2Cparameter-schema#pipeline-structure) parameter with one of our [Cypress Docker Images](https://github.com/cypress-io/cypress-docker-images/tree/master/browsers#readme) will ensure that all of the workers have the same version of Node along with browsers.

**Comment by Rutix** [comment]
@kevinold That sounds like a good plan! We are gonna try to see if we can use a Cypress Docker image and will report back! Thanks for the tip.
